### PR TITLE
f32: stop clobbering R14 (goroutine g) in realFFTUnpackAVX

### LINKED ISSUE #58: amd64: realFFTUnpackAVX uses R14 (reserved for goroutine g) as a scratch register
## Summary

`realFFTUnpackAVX` in `f32/f32_amd64.s` (roughly lines 4037-4207) uses `R14` as a
general-purpose scratch register for byte-offset and pointer math (`MOVQ CX, R14`,
`SHLQ`, `ADDQ R14, ...`, etc.). On Go's amd64 internal ABI, `R14` is reserved for
the current goroutine pointer `g`
([abi-internal.md](https://go.googlesource.com/go/+/refs/heads/master/src/cmd/compile/abi-internal.md)).

This is the same register-hygiene issue that was fixed for the ConvolveDecimate
kernels in #57 (there the strided `pos` was moved from `R14` to `BX`).

## Not an active crash (verified)

`realFFTUnpackAVX` is a leaf `NOSPLIT` function with no `CALL` into the runtime,
so it does not crash today:

- the ABIInternal->ABI0 call wrapper restores `g` into `R14` from TLS on return, and
- stackmap-less hand-written assembly is never async-preempted, so the preemption
  handler never reads the clobbered `R14` mid-function.

Both were confirmed empirically on go1.26 while working on #57 (a 2M-iteration
alloc loop and a 5000x2M long-spin loop under forced GC, neither crashed).

So this is **hardening / consistency**, not a live bug. The risk is latent: the
moment this function (or a copy of it) grows a `CALL` into the runtime, or gains a
stack map, the clobbered `g` becomes a GC-time crash.

## Suggested fix

Move the `R14` scratch usage to a non-reserved register. `R12`/`R13` are permanent
scratch; legacy regs (`BX`, etc.) also avoid the REX prefix. Pick whatever is free
in the function. After the change, `grep -n R14 f32/f32_amd64.s` should show no
instruction-level R14 use on amd64 (only the convention note in comments).

The reserved-register convention is now documented in `CLAUDE.md`.

## Notes

- ARM64 `R14` is unaffected: on arm64 `g` is `R28`, and `R14` is an ordinary
  argument/result register, so the NEON code's `R14` use is fine.
- `f64/f64_amd64.s` has no instruction-level R14 use after #57.

Low priority.

### PR DESCRIPTION
**Pull request by tphakala**
## Summary

`realFFTUnpackAVX` in `f32/f32_amd64.s` used `R14` as a scratch register for byte-offset and pointer math. On Go's amd64 internal ABI `R14` holds the current goroutine pointer `g`, so clobbering it is a latent GC-time crash the moment the function grows a `CALL` into the runtime or gains a stack map. This is the same hazard that was fixed for the ConvolveDecimate kernels in #57, and the convention is documented in `CLAUDE.md`.

- Moved all `R14` scratch usage in `realFFTUnpackAVX` to `BX`: unused elsewhere in the function, non-reserved, and REX-free. Pure register rename, no behavioral change.
- Added `TestNoGoroutineRegisterClobber`, a regression guard that walks every `.s` file and fails if the goroutine-`g` register is used as an instruction operand (`R14` on amd64, `R28`/`W28` on arm64). Line comments are stripped first, so the convention notes that name the register for documentation stay legal. `R15` is intentionally not flagged since it is sanctioned static scratch in this repo.

`grep` now shows no instruction-level `R14` in `f32/f32_amd64.s`, only the convention-note comments.

## Related Issues

Fixes #58

## Test Plan

- [x] `go test ./...` green on amd64 (native) and arm64 (native, Raspberry Pi 5)
- [x] `RealFFTUnpack` parity / known-value / edge-case tests confirm bit-identical output
- [x] `asmcheck` WORD cross-validation green on both architectures
- [x] New regression test proven RED on the pre-fix tree (flags all 25 `R14` sites, ignores comment references) and GREEN after the fix
- [x] `go vet` clean on the changed files, `gofmt` clean, cross-compile builds clean for amd64 and arm64

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed a critical assembly code bug where improper register usage during floating-point scalar operations could cause runtime failures across supported platforms.

* **Tests**
  * Added automated regression test that validates assembly code across all supported architectures, preventing similar register handling violations.

### CODE REVIEW

File: asmcheck_test.go [hunk 1]
```diff
@@ -202,3 +203,74 @@ func TestNoUncheckedAmd64Encodings(t *testing.T) {
...
+	}
+
+	var violations []string
+	err := filepath.WalkDir(".", func(p string, d os.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() || !strings.HasSuffix(p, ".s") {
+			return nil
+		}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply asmcheck_test.go:241]: Done in 641da72. Normalized the path with filepath.ToSlash at the WalkDir boundary (matching findArm64Asm) and simplified the reporting site to plain p.

**Comment by tphakala** [comment]
@coderabbitai good catch on the dead X14 leftover. Removed it in 641da72, along with two adjacent dead blocks in the same function: the superseded R9 reverse-pointer first attempt (the "Wrong, redo this" lines) and the unused Y15 abs mask. I kept lines 4309-4311 (the 0.5 broadcast into X13/Y13) since that register is live and used throughout the loop. RealFFTUnpack parity and known-value tests still pass bit-identical on amd64 and arm64.

@coderabbitai review

File: asmcheck_test.go [hunk 2]
```diff
@@ -202,3 +203,77 @@ func TestNoUncheckedAmd64Encodings(t *testing.T) {
...
+// static scratch in this repo (see CLAUDE.md), unlike g.
+func TestNoGoroutineRegisterClobber(t *testing.T) {
+	// g lives in R14 on amd64 and R28 on arm64 (W28 is the 32-bit view; a
+	// write to it zero-extends over g all the same).
+	gReg := []struct {
+		suffix string
+		name   string
+		re     *regexp.Regexp
+	}{
+		{"_amd64.s", "R14", regexp.MustCompile(`\bR14\b`)},
```
[the remark this answers is not in the document]
Comment (tphakala) [reply asmcheck_test.go:227]: Good instinct on the sub-register hazard, but I verified this against the Go assembler and the current regex already covers it. The Intel-style spellings R14D/R14W/R14B are not valid Plan 9 asm: Go rejects `MOVL $0, R14B` as `invalid instruction`. Go expresses a 32/16/8-bit write to g as a sized mnemonic on the same R14 token, e.g. `MOVL $0, R14`, which does zero-extend and clobber g, exactly the case you flagged. Since the operand is still spelled R14, `\bR14\b` catches it; the RED run of this test in fact flagged the existing `MOVL $0x3F000000, R14` lines. arm64 is covered the same way: a 32-bit write is `W28`, already matched by `\b[RW]28\b`. So I am keeping `\bR14\b`.
